esmvalcore/preprocessor/_area.py

```diff
@@ -243,33 +272,60 @@ def area_statistics(cube, operator, fx_files=None):
     ValueError
         if input data cube has different shape than grid area weights
     """
-    grid_areas = tile_grid_areas(cube, fx_files)
+    if operator == 'weightless_sum':
+        return cube.collapsed(['longitude', 'latitude'], iris.analysis.SUM)
+
+    if operator == 'weightless_mean':
+        return cube.collapsed(['longitude', 'latitude'], iris.analysis.MEAN)
+
+    operation = get_iris_analysis_operation(operator)
 
+    # Many IRIS analysis functions do not accept weights arguments.
+    if not operator_accept_weights(operator):
+        return cube.collapsed(['longitude', 'latitude'], operation)
+
+    # The following are for operators with means
     if not fx_files and cube.coord('latitude').points.ndim == 2:
         logger.error(
-            'fx_file needed to calculate grid cell area for irregular grids.')
+            'fx_file needed for weighted calcuions on irregular grids.')
         raise iris.exceptions.CoordinateMultiDimError(cube.coord('latitude'))
 
-    coord_names = ['longitude', 'latitude']
-    if grid_areas is None or not grid_areas.any():
-        cube = guess_bounds(cube, coord_names)
+    if fx_files and calculate_grid:
+        logger.error(
+            'Cannot calculate_grid when FX file is provided.')
+        raise ValueError('Incompatible recipe arguments (area_statistics): '
+                         'fx_files and calculate_grid do not work together.')
+
+    # Load grid area
+    grid_areas_loaded = False
+    if fx_files:
+        grid_areas = load_fx(fx_files)
+        grid_areas_loaded = True
+
+    if calculate_grid:
+        cube = guess_bounds(cube, ['longitude', 'latitude'])
         grid_areas = iris.analysis.cartography.area_weights(cube)
         logger.info('Calculated grid area shape: %s', grid_areas.shape)
+        grid_areas_loaded = True
+
+    if not grid_areas_loaded:
+        logger.error(
+            'Cannot calculate_grid without Grid area data. '
+            'Please provide fx_files or set calculate_grid: True.')
+        raise ValueError('Cannot calculate_grid without Grid area data.')
+
+    # Tile grid area to match cube
+    grid_areas = tile_grid_areas(cube, grid_areas)
 
     if cube.shape != grid_areas.shape:
         raise ValueError('Cube shape ({}) doesn`t match grid area shape '
                          '({})'.format(cube.shape, grid_areas.shape))
 
-    operation = get_iris_analysis_operation(operator)
-
-    # TODO: implement weighted stdev, median, s var when available in iris.
-    # See iris issue: https://github.com/SciTools/iris/issues/3208
+    if isinstance(grid_areas, da.Array):
+        grid_areas = grid_areas.compute()
```

Comment:
Why are you realizing the data? Is it not possible to use a dask array as a weight?

Reply:
Nope, iris only takes numpy arrays as weights.